src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -190,14 +212,24 @@ impl RxBuffers {
 
     /// This will let the guest know that about all the `DescriptorChain` object that has been
     /// used to receive a frame from the TAP.
-    fn finish_frame(&mut self, dc: &ParsedDescriptorChain, rx_queue: &mut Queue) {
-        // It is fine to `.unrap()` here. The only reason why `add_used` can fail is if the
-        // `head_index` is not a valid descriptor id. `head_index` here is a valid
-        // `DescriptorChain` index. We got it from `queue.pop_or_enable_notification()` which
-        // checks for its validity. In other words, if this unwrap() fails there's a bug in our
-        // emulation logic which, most likely, we can't recover from. So, let's crash here
-        // instead of logging an error and continuing.
-        rx_queue.add_used(dc.head_index, dc.length).unwrap();
+    fn finish_frame(&mut self, rx_queue: &mut Queue) {
+        rx_queue.advance_used_ring(self.used_descriptors);
+        self.used_descriptors = 0;
+        self.used_bytes = 0;
+    }
+
+    /// Return a slice of iovecs for the first slice in the buffer.
+    ///
+    /// # Safety
+    /// Buffer needs to have at least one descriptor chain parsed.
+    unsafe fn single_chain_slice_mut(&mut self) -> &mut [iovec] {
+        let nr_iovecs = self.parsed_descriptors[0].nr_iovecs as usize;
+        &mut self.iovec.as_iovec_mut_slice()[..nr_iovecs]
+    }
```

Comment:
I don't think this needs to be `unsafe` (same for `read_tap` actually), because it just panics if there's no descriptors, right?

Reply:
If we follow this logic, we can make any function `safe` if we just assert all invariants in the function itself. I don't have strong opinions here. I can remove this if you want.

Reply:
> If we follow this logic, we can make any function safe if we just assert all invariants in the function itself.
Yes, we could :D That's pretty much the difference between `std::str::from_utf8` and `std::str::from_utf8_unchecked` - omitting assertions/checks for performance, at the cost of an `unsafe`. Its more difficult with functions like `VolatileSlice::new(ptr: *mut u8, len: usize)` - there's no assertion you can write that would guarantee that `ptr` points to valid memory of length `len`, so the function must be unsafe in every incarnation.

Reply:
So, do we leave this as it is or remove `unsafe` qualifier? Actually if we keep them `unsafe` we can do:
```Rust
let nr_iovecs = self.parsed_descriptors.get_unchecked(0).nr_iovecs as usize;
```

Reply:
`VecDeque` doesn't have a `get_unchecked`, so let's go with removing `unsafe` here

Reply:
done